apps/ssh/ssh-add2.1
.\"  -*- nroff -*-
.\"
.\" ssh-add2.1
.\"
.\" Author:     Tatu Ylonen <ylo@ssh.fi>
.\"             Markku-Juhani Saarinen <mjos@ssh.fi>
.\"
.\" Copyright (c) 1997 SSH Communications Security, Finland
.\" All rights reserved
.\"
.nr CO 0
.ie \n(CO .TH SSH-ADD2 1 "May 30, 1998" "SSH 2" "SSH 2"
.el .TH SSH-ADD2 1 "May 30, 1998" "SSH2" "SSH2"

.SH NAME
ssh-add2 \- adds identities for the authentication agent

.SH SYNOPSIS
.B ssh-add2
[\c
.B \-p\
]
[\c
.B \-l\
]
[\c
.B \-d\c
]
[\c
.B \-D\c
]
[\c
.B \-L\c
]
[\c
.B \-U\c
]
[\c
.B \-1\c
]
[\c
.B \-u\c
]
[\c
.B \-f forwarding steps\c
]
[\c
.B \-F forwarding constraint\c
]
[\c
.B \-t key timeout in minutes\c
]
[\c
.B files\c
\&.\|.\|.\|]

.SH DESCRIPTION 
.LP
.B ssh-add2
adds identities to the authentication agent,
.B ssh-agent2.
If any file requires a passphrase,
.B ssh-add2
asks for the passphrase from the user.  If the -p option is given then
the passphrase is read from stdin, otherwise if the user is using X11,
the passphrase is requested using a small X11 program; otherwise it is
read from the user's tty.  (Note: it may be necessary to redirect stdin
from /dev/null to get the passphrase requested using X11.)
.LP
The authentication agent must be running and must be an ancestor of
the current process for
.B ssh-add2
to work.

If
.B ssh-add2
needs a passphrase, it will read the passphrase from the current
terminal if it was run from a terminal.  If
.B ssh-add2
does not have a terminal associated with it but
.SM DISPLAY\s0
is set, it
will open an X11 window to read the passphrase.  This is particularly
useful when calling
.B ssh-add2
from a .Xsession or related script.  (Note that on some machines it
may be necessary to redirect the input from /dev/null to make this work.)


.SH OPTIONS
.TP 0.5i
.B \-p
Read passphrase from stdin (or pipe).
.TP
.B \-l
Lists all identities currently represented by the agent.
.TP
.B \-d
Instead of adding the identity, removes the identity from the agent.
.TP
.B \-D
Deletes all identities from the agent.
.TP
.B \-L
Temporarily locks the agent with a password.
.TP
.B \-U
Unlocks the locked agent.
.TP
.B \-1
Agent is not allowed to use keys added with this command
in ssh1 compatibility operations.	
.TP
.B \-u
The keys added are not read from the file, but instead the key is 
given to the agent as URL.  With this functionality, agent can get
information about the additional key sources, like smartcards.
.TP
.B \-f forwarding steps
Key can be used only through as many forwarding steps as directed by
the argument.  Argument 0 states that key can only be used locally.
Be aware, that ssh1 does not submit forwarding information, so with
ssh1 compatibility, this constraint may not work as expected.
.TP
.B \-F forwarding constraint
The argument is a comma separated list, through which kind of steps
the key can be forwarded.  For example constraint string
"*.ssh.fi,rinne.iki.fi" 
states that key can be forwarded to any host in domain 
.I ssh.fi
and also to host
.I rinne.iki.fi
and used locally.
Be aware, that ssh1 does not submit forwarding information, so with
ssh1 compatibility, this constraint may not work as expected.
.TP
.B \-t timeout
The agent is advised to delete the key after timeout.  Timeout is
given in minutes.

.SH RETURN STATUS

.B ssh-add2
returns one of the following exit statuses.  These may be useful in scripts.
.IP 0
The requested operation was performed successfully.
.IP 1
No connection could be made to the authentication agent.  Presumably
there is no authentication agent active in the execution environment
of
.B ssh-add2.
.IP 2
The user did not supply a required passphrase.
.IP 3
An identify file could not be found, was not readable, or was in
bad format.
.IP 4
The agent does not have the requested identity.
.IP 5
An unspecified error has occurred; this is a catch-all for errors not
listed above.
.RT

.SH FILES
.TP 
.I \&$HOME/\s+2.\s0ssh2/identification
Contains names of the private keys that are to be used in authentication.
See 
.BR ssh2 (1)
for more information.
.TP
.I \&$HOME/\s+2.\s0ssh2/id_KEYTYPE_KEYLEN_X
.TP
.I \&$HOME/\s+2.\s0ssh2/id_KEYTYPE_KEYLEN_X.pub
Standard private and public identification key files.


.SH AUTHORS
.LP

SSH Communications Security Oy

For more information, see http://www.ssh.fi.

.SH SEE ALSO
.BR ssh-agent2 (1),
.BR ssh-keygen2 (1),
.BR ssh2 (1),
.BR sshd2 (8)


CHANGES
1999-01-29  Sami Lehtinen  <sjl@ssh.fi>

	* ssh-2.0.12.

	* Fixed a deficiency in the configure script. Because of it the
 	  sp_expire and sp_inact fields of shadow-password-struct were
 	  overlooked.

1999-01-27  Sami Lehtinen  <sjl@ssh.fi>

	* Added code to trcommon.c/ssh_tr_input_kex2 to check for return
	  value of tr->kex->{server,client}_input_kex2, and to send
	  disconnect if return value was FALSE. Should make error messages
	  in certain situations more clear.

	* Added code to sshconn.c to check if we receive EOF from the main
	  stream. Didn't seem to brake anything, and should fix some hanging
	  problems.

	* Fixed a bug in lib/sshutil/sshtcp.c. Previously, it ignored
	  definition for port with SSH_SOCKS_SERVER.

	* Fixed check for broken inet_ntoa, and it now seems to work (it
	  compiled and worked ok on mips-sgi-irix6.2 with gcc 2.8.1, where
	  it didn't before).

1999-01-18  Sami Lehtinen  <sjl@ssh.fi>

	* ssh-2.0.12.pre1 (ie. pre-release 1).

	* Changed ssh2 so that if we are root, it won't fetch keys and
	  config from /etc/ssh2, but from $HOME/.ssh2. 

	* Added code to configure to better chack for minor (is a macro in
	  atleast Solaris) and S_IFSOCK (isn't defined in SCO). Also added
	  code to check for broken inet_ntoa, but it doesn't seem to work
	  as expected. Studies continue.

	* Fixed a bug in ssh_user_dir() and ssh2.c , which caused ssh2 to
 	  crash when run on the first time in some systems.

	* Added FAQ. Send me good questions, and (more importantly) good
	  answers, and I will put them to the FAQ.

	* Added signal handlers for various fatal signals in sftp2 and in
 	  scp2. If we receive a fatal signal, we will also kill the
 	  ssh2-child. Fixed a bug in filecontrol (or rather added a
 	  kludge). We set stdio and stdout to blocking mode now in main(),
 	  as they are somewhere (eventloop_initialize ?) put to non-blocking
 	  mode. This broke output in, for example, the "ls" command. Also
 	  rewrote bits of code, sftp_page_prompt_return for example.

1999-01-17  Timo J. Rinne  <tri@ssh.fi>

	* Added configuration option --with(out)-ssh-agent1-compat
	  to make ssh-agent1 support (RSA-decryption) optional
	  in ssh-agent2.  Default is on.

1999-01-12  Timo J. Rinne  <tri@ssh.fi>

	* Added ssh1 challenge-response mechanism into ssh-agent2.

1998-12-29  Sami Lehtinen  <sjl@ssh.fi>

	* Added configuration parameter SyslogFacility. Now you can define
	  exactly where you want to log things. Sorry it took so long.

	* Fixed a security bug which allowed  any eligible user to request
	  remote forwarding from privileged ports without being root. 

	* Previous: Modified the configuration script so that it should
	  compile on HP-UX 9.x now.

1998-12-10  Sami Lehtinen  <sjl@ssh.fi>

	* Added {Allow,Deny}Hosts configuration parameters to
	  sshd2. Accepts wildcards (*, ?). Will accept in the future:
	  address ranges, subnets. Also a access-control-file is planned,
	  which would make this obsolete, as the new format would be much
	  more powerful and intuitive.

1998-12-08  Sami Lehtinen  <sjl@ssh.fi>

	* Added SIGHUP handling to sshd2. (now restarts on SIGHUP)

1998-11-26  Sami Lehtinen  <sjl@ssh.fi>

	* Tweaked logging in sshd2.c.

	* Added reverse-mapping for the remote host's hostname in the
 	  server.

	* Fixed a bug in auths-{passwd,pubkey}.c, which caused a SIGSEGV
 	  if user didn't exist. Tweaked logging in same file.

1998-11-16  Sami Lehtinen  <sjl@ssh.fi>

	* Released  ssh-2.0.11.
	
	* Previous: Fixed a draft inconsistency in the client's version
 	  string handling. Should reduce those "Protocol error"'s.

	* Hopefully fixed the busyloop-timeout problem in scp2. Now it
	  kills the background ssh2 process before exiting itself.

1998-11-12  Timo J. Rinne  <tri@ssh.fi>

	* Added twofish encryption.

1998-11-12  Sami Lehtinen  <sjl@ssh.fi>

	* Previous: the #include bug in sshreadline.c, which manifested
	  itself in solaris 2.[456].x, should now be fixed. (atleast in our
	  server compiles ok).

	* Added --disable-asm flag to configure. If asm-optimizations
	  don't work, use this.

	* Added code to configure to properly add libipc for
	  bsdi2.1. In AIX 4.* utmpx is now disabled (utmpx.h is reportedly
	  broken). With HPUX _HPUX_SOURCE is now properly defined.

	* Changed handling of setsid-errors in ssh2 and sshd2. Previously
	  if setsid() call failed, ssh{2,d2} called ssh_fatal. Now, it gives
	  warning in ssh2 and logs the event in sshd2. This is the same
	  behavior as in ssh1.

	* Implemented '-r'-flag (copy directories recursively) in
	  scp2. Also cleaned up code in scp2.

1998-11-05  Sami Lehtinen  <sjl@ssh.fi>

	* Added '-1' flag, which enables scp1 compatibility. Use it as the
 	  first argument to scp2. Implemented wildcard-expansion
 	  (globbing). Currently supported wildcards are '?' and '*'.

1998-10-29  Sami Lehtinen  <sjl@ssh.fi>

	* Made progress-indicator default in scp2. Added some code to
	  display transfer-times and transfer-speed.

1998-10-29  Timo J. Rinne  <tri@ssh.fi>

	* Ssh-agent2 now more or less fully supports also requests
	  sent by ssh1 and ssh-add1. To enable ssh1 compatibility
	  in ssh-agent2 you have to start it with option -1.
	  CAUTION: ssh-agent2 works properly only with versions
	  ssh-2.0.11 and above, if run with -1 option.  Without
	  -1 option it's compatible with earlier ssh-2.0.* versions
	  too.

1998-10-21  Timo J. Rinne  <tri@ssh.fi>

	* Moved utmp update to the child.  It now seems to work
	  at least in BSD.

1998-10-19  Sami Lehtinen  <sjl@ssh.fi>

	* Added correct parsing for specifying 'user@host' on the
	  commandline. (`ssh2 user@host' equals `ssh2 -luser host')

	* Fixed --with-etcdir and --with-libwrap options in configure and
	  apps/ssh/Makefile.am.

	* Fixed calculating MAC, as it was done against the draft. Added
	  compatibility code, so that we can work with older
	  ssh-2.0.x-versions (2.0.[789]) still.

	* Implemented PasswordAuthentication and PubkeyAuthentication

apps/ssh/ssh-add2.c
/*

ssh-add.c

  Authors:
        Tatu Ylonen <ylo@ssh.fi>
        Markku-Juhani Saarinen <mjos@ssh.fi>
        Timo J. Rinne <tri@ssh.fi>
        Sami Lehtinen <sjl@ssh.fi>

  Copyright (C) 1997-1998 SSH Communications Security Oy, Espoo, Finland
  All rights reserved.

Adds an identity to the authentication server, or removes an identity.

*/

/*
 * $Id: ssh-add2.c,v 1.18 1999/01/20 12:30:44 tri Exp $
 * $Log: ssh-add2.c,v $
 * $EndLog$
 */

#include "sshincludes.h"
#include "sshcrypt.h"
#include "sshtimeouts.h"
#include "sshagent.h"
#include "sshuser.h"
#include "readpass.h"
#include "sshuserfiles.h"
#include "sshunixeloop.h"
#include "sshgetopt.h"

#define SSH_DEBUG_MODULE "SshAdd"

#ifdef HAVE_LIBWRAP
int allow_severity = SSH_LOG_INFORMATIONAL;
int deny_severity = SSH_LOG_WARNING;
#endif /* HAVE_LIBWRAP */

#define EXIT_STATUS_OK          0
#define EXIT_STATUS_NOAGENT     1
#define EXIT_STATUS_BADPASS     2
#define EXIT_STATUS_NOFILE      3
#define EXIT_STATUS_NOIDENTITY  4
#define EXIT_STATUS_ERROR       5

typedef enum 
{ 
  LIST, 
  ADD, 
  ADD_URL,
  DELETE,
  DELETE_URL,
  DELETE_ALL, 
  LOCK, 
  UNLOCK 
} SshAgentAction;

/* Files to add/delete from agent. */
char **files;
int num_files;
SshAgentAction action;

/* Possible attributes */
SshUInt32 path_limit = 0xffffffff;
char *path_constraint = NULL;
Boolean forbid_compat = FALSE;
time_t key_timeout = 0;
Boolean have_attrs = FALSE;
SshUInt32 use_limit = 0xffffffff;

/* Force to read passphrases from stdin. */
int use_stdin = FALSE;

/* Information about the current user. */
SshUser user;

void agent_completion(SshAgentError result, void *context);

void add_file(SshAgent agent, const char *filename)
{
  SshPrivateKey key;
  char *saved_comment, *comment, *pass;
  int query_cnt;
  unsigned char *certs;
  size_t certs_len;
  char privname[500], pubname[500];
  unsigned long magic;
  struct stat st;

  if (action == ADD_URL)
    {
      printf("Adding URL identity: %s\n", filename);
      snprintf(privname, sizeof(privname), "%s", filename);
      if (have_attrs)
        ssh_agent_add_with_attrs(agent, NULL, NULL, 0, privname,
                                 path_limit, path_constraint, use_limit,
                                 forbid_compat, key_timeout,
                                 agent_completion, (void *)agent);
      else
        ssh_agent_add(agent, NULL, NULL, 0, privname,
                      agent_completion, (void *)agent);
      return;
    }
  else if (action == DELETE_URL)
    {
      printf("Deleting URL identity: %s\n", filename);
      snprintf(privname, sizeof(privname), "%s", filename);
      ssh_agent_delete(agent, NULL, 0, privname,
                       agent_completion, (void *)agent);
      return;
    }
  /* Construct the names of the public and private key files. */
  if (strlen(filename) > 4 &&
      strcmp(filename + strlen(filename) - 4, ".pub") == 0)
    {
      snprintf(pubname, sizeof(pubname), "%s", filename);
      snprintf(privname, sizeof(privname), "%s", filename);
      privname[strlen(privname) - 4] = '\0';
    }
  else
    {
      snprintf(pubname, sizeof(pubname), "%s.pub", filename);
      snprintf(privname, sizeof(privname), "%s", filename);
    }

  if (action == ADD)
    printf("Adding identity: %s\n", pubname);
  else if (action == DELETE)
    printf("Deleting identity: %s\n", pubname);

  if (stat(pubname, &st) < 0)
    {
      printf("Public key file %s does not exist.\n", pubname);
      (*agent_completion)(SSH_AGENT_ERROR_OK, (void *)agent);
      return;
    }

  if (stat(privname, &st) < 0)
    {
      printf("Private key file %s does not exist.\n", privname);
      (*agent_completion)(SSH_AGENT_ERROR_OK, (void *)agent);
      return;
    }
  
  /* Read the public key blob. */
  magic = ssh2_key_blob_read(user, pubname, &saved_comment, &certs, &certs_len,
                            NULL);
  if (magic != SSH_KEY_MAGIC_PUBLIC)
    {
      printf("Bad public key file %s\n", pubname);
      (*agent_completion)(SSH_AGENT_ERROR_OK, (void *)agent);
      return;
    }
  
  if (action == ADD)
    {
      /* Loop until we manage to load the file, or a maximum number of
         attempts have been made.  First try with an empty passphrase. */
      pass = ssh_xstrdup("");
      query_cnt = 0;
      while ((key = ssh_privkey_read(user, privname, pass, &comment, NULL)) == 
             NULL)
        {
          char buf[1024];
          FILE *f;
      
          /* Free the old passphrase. */
          memset(pass, 0, strlen(pass));
          ssh_xfree(pass);
          
          query_cnt++;
          if (query_cnt > 5)
            {
              fprintf(stderr,
                      "You don't seem to know the correct passphrase.\n");
              exit(EXIT_STATUS_BADPASS);
            }
          
          /* Ask for a passphrase. */
          if (!use_stdin && getenv("DISPLAY") && !isatty(fileno(stdin)))
            {
              snprintf(buf, sizeof(buf),
                       "ssh-askpass2 '%sEnter passphrase for %.100s'", 
                       ((query_cnt <= 1) ? 
                        "" : 
                        "You entered wrong passphrase.  "), 
                       saved_comment);
              f = popen(buf, "r");
              if (!fgets(buf, sizeof(buf), f))
                {
                  pclose(f);
                  ssh_xfree(saved_comment);
                  exit(EXIT_STATUS_BADPASS);
                }
              pclose(f);
              if (strchr(buf, '\n'))
                *strchr(buf, '\n') = 0;
              pass = ssh_xstrdup(buf);
            }
          else
            {
              if (query_cnt <= 1)
                printf("Need passphrase for %s (%s).\n", 
                       privname, saved_comment);
              else
                printf("Bad passphrase.\n");
              pass = ssh_read_passphrase("Enter passphrase: ", use_stdin);
              if (pass == NULL || strcmp(pass, "") == 0)
                {
                  ssh_xfree(saved_comment);
                  ssh_xfree(pass);
                  exit(EXIT_STATUS_BADPASS);
                }
            }
        }
      memset(pass, 0, strlen(pass));
      ssh_xfree(pass);
      ssh_xfree(saved_comment);
      /* Construct a comment for the key by combining file name and comment in
         the file. */
      snprintf(privname, sizeof(privname), "%s: %s", pubname, comment);
      ssh_xfree(comment);
    }
  else
    {
      /* Construct a comment for the key by combining file name and comment in
         the file. */
      snprintf(privname, sizeof(privname), "%s: %s", pubname, saved_comment);
      ssh_xfree(saved_comment);
    }

  if (action == ADD)
    {
      /* Send the key to the authentication agent. */
      if (have_attrs)
        ssh_agent_add_with_attrs(agent, key, certs, certs_len, privname,
                                 path_limit, path_constraint, use_limit,
                                 forbid_compat, key_timeout,
                                 agent_completion, (void *)agent);
      else
        ssh_agent_add(agent, key, certs, certs_len, privname,
                      agent_completion, (void *)agent);
      ssh_private_key_free(key);
    }
  else if (action == DELETE)
    {
      ssh_agent_delete(agent, certs, certs_len, privname,
                       agent_completion, (void *)agent);
    }
}

void agent_completion(SshAgentError result, void *context)
{
  SshAgent agent = (SshAgent)context;

  switch (result)
    {
    case SSH_AGENT_ERROR_OK:
      break;

    case SSH_AGENT_ERROR_TIMEOUT:
      fprintf(stderr, "Authentication agent timed out.\n");
      exit(EXIT_STATUS_NOAGENT);
      
    case SSH_AGENT_ERROR_KEY_NOT_FOUND:
      fprintf(stderr,
              "Requested key not in possession of authentication agent.\n");
      exit(EXIT_STATUS_NOIDENTITY);
      
    case SSH_AGENT_ERROR_DECRYPT_FAILED:
      fprintf(stderr, "Decryption failed.\n");
      exit(EXIT_STATUS_ERROR);
      
    case SSH_AGENT_ERROR_SIZE_ERROR:
      fprintf(stderr, "Argument size error.\n");
      exit(EXIT_STATUS_ERROR);
      
    case SSH_AGENT_ERROR_KEY_NOT_SUITABLE:
      fprintf(stderr, 
              "The specified key is not suitable for the operation.\n");
      exit(EXIT_STATUS_ERROR);
      
    case SSH_AGENT_ERROR_DENIED:
      fprintf(stderr, "The requested operation was denied.\n");
      exit(EXIT_STATUS_ERROR);
      
    case SSH_AGENT_ERROR_FAILURE:
      fprintf(stderr, "The requested operation failed.\n");
      exit(EXIT_STATUS_ERROR);
      
    case SSH_AGENT_ERROR_UNSUPPORTED_OP:
      fprintf(stderr, "The requested operation is not supported.\n");
      exit(EXIT_STATUS_ERROR);
      
    case SSH_AGENT_ERROR_BUSY:
      fprintf(stderr, "The authentication agent is busy.\n");
      exit(EXIT_STATUS_ERROR);
      
    default:
      fprintf(stderr, "Authentication agent failed with error %d\n",
              (int)result);
      exit(EXIT_STATUS_ERROR);
    }

  /* The last operation was successful.  Check if there is more work to do. */
  if (num_files <= 0)
    {
      /* All operations performed. */
      exit(EXIT_STATUS_OK);
    }

  /* Add any files listed. */
  num_files--;
  add_file(agent, *files++);
  /* A callback should already have been scheduled to occur at some point. */
}

void agent_list_callback(SshAgentError error, unsigned int num_keys,
                             SshAgentKeyInfo keys, void *context)
{
  SshAgent agent = (SshAgent)context;
  int i;

  if (error != SSH_AGENT_ERROR_OK)
    {
      agent_completion(error, (void *)agent);
      ssh_fatal("agent_list_callback: agent_completion returned after error");
    }

  if (num_keys == 0)
    printf("The authorization agent has no keys.\n");
  else
    {
      if (num_keys == 1)
        printf("The authorization agent has one key:\n");
      else
        printf("The authorization agent has %d keys:\n", num_keys);
      for (i = 0; i < num_keys; i++)
        printf("%s\n", keys[i].description);
    }
  agent_completion(SSH_AGENT_ERROR_OK, (void *)agent);
}

void agent_open_callback(SshAgent agent, void *context)
{
  char buf[1024];
  FILE *f;
  char *password, *password_vrfy;

  if (!agent)
    {
      fprintf(stderr,
        "Failed to connect to authentication agent - agent not running?\n");
      exit(EXIT_STATUS_NOAGENT);
    }

  switch (action)
    {
    case DELETE_ALL:
      fprintf(stderr, "Deleting all identities.\n");
      ssh_agent_delete_all(agent, agent_completion, (void *)agent);
      break;
      
    case LIST:
      fprintf(stderr, "Listing identities.\n");
      ssh_agent_list(agent, agent_list_callback, (void *)agent);
      break;
      
    case LOCK:
      if (!use_stdin && getenv("DISPLAY") && !isatty(fileno(stdin)))
        {
          snprintf(buf, sizeof (buf),
                   "ssh-askpass2 'Enter lock passphrase'");
          f = popen(buf, "r");
          if (! fgets(buf, sizeof (buf), f))
            {
              pclose(f);
              exit(EXIT_STATUS_BADPASS);
            }
          pclose(f);
          if (strchr(buf, '\n'))
            *strchr(buf, '\n') = 0;
          password = ssh_xstrdup(buf);
        }
      else
        {
          password = ssh_read_passphrase("Enter lock password: ", use_stdin);
          if (password == NULL)
            {
              fprintf(stderr, "Unable to read password.\n");
                exit(EXIT_STATUS_BADPASS);
            }
          password_vrfy = ssh_read_passphrase("Again: ", use_stdin);
          if (password_vrfy == NULL)
            {
              fprintf(stderr, "Unable to read password.\n");
              exit(EXIT_STATUS_BADPASS);
            }
          if (strcmp(password, password_vrfy) != 0)
            {
              fprintf(stderr, "Passwords don't match.\n");
              exit(EXIT_STATUS_BADPASS);
            }
        }
      ssh_agent_lock(agent, password, agent_completion, (void *)agent);
      break;

    case UNLOCK:
      if (!use_stdin && getenv("DISPLAY") && !isatty(fileno(stdin)))
        {
          snprintf(buf, sizeof (buf),
                   "ssh-askpass2 'Enter unlock passphrase'");
          f = popen(buf, "r");
          if (! fgets(buf, sizeof (buf), f))
            {
              pclose(f);
              exit(EXIT_STATUS_BADPASS);
            }
          pclose(f);
          if (strchr(buf, '\n'))
            *strchr(buf, '\n') = 0;
          password = ssh_xstrdup(buf);
        }
      else
        {
          password = ssh_read_passphrase("Enter lock password: ", use_stdin);
          if (password == NULL)
            {
              exit(EXIT_STATUS_BADPASS);
            }
        }
      ssh_agent_unlock(agent, password, agent_completion, (void *)agent);
      break;

    case DELETE:
    case DELETE_URL:
    case ADD:
    case ADD_URL:
      /* Let the completion do all the work. */
      agent_completion(SSH_AGENT_ERROR_OK, (void *)agent);
      break;

    default:
      ssh_fatal("agent_open_callback: bad action %d\n", (int)action);
    }
}

void usage(void);
void usage()
{
  fprintf(stderr, "Usage: ssh-add [-l] [-d] [-D] [-p] [-t key_exp] [-f hop_limit] [-1] [files...]\n");
}

/* This is the main program for the agent. */

int main(int ac, char **av)
{
  int opt, i, len;
  DIR *ssh2dir = NULL;
  char *ssh2dirname;
  Boolean dynamic_array = FALSE;
  struct dirent * cand;

  user = ssh_user_initialize(NULL, FALSE);
  
  action = ADD;
  while ((opt = ssh_getopt(ac, av, "ldDput:f:F:1LU", NULL)) != EOF)
    {
      if (!ssh_optval)
        {
          usage();
          exit(EXIT_STATUS_ERROR);
        }
      switch (opt)
        {
        case 'l':
          action = LIST;
          break;
        case 'p':
          use_stdin = TRUE;
          break;
        case 'd':
          if (action == ADD_URL)
            action = DELETE_URL;
          else
            action = DELETE;
          break;
        case 'D':
          action = DELETE_ALL;
          break;
        case 't':
          if (ssh_optargnum)
            {
              key_timeout = (time_t)(ssh_optargval * 60);
            }
          else
            {
              usage();
              exit(EXIT_STATUS_ERROR);
            }
          have_attrs = TRUE;
          break;
        case 'f':
          if (ssh_optargnum)
            {
              path_limit = (SshUInt32)ssh_optargval;
            }
          else
            {
              usage();
              exit(EXIT_STATUS_ERROR);
            }
          have_attrs = TRUE;
          break;
        case 'F':
          path_constraint = ssh_xstrdup(ssh_optarg);
          have_attrs = TRUE;
          break;
        case '1':
          forbid_compat = TRUE;
          have_attrs = TRUE;
          break;
        case 'u':
          if (action == DELETE)
            action = DELETE_URL;
          else
            action = ADD_URL;
          break;
        case 'L':
          action = LOCK;
          break;
        case 'U':
          action = UNLOCK;
          break;
        default:
          usage();
          exit(EXIT_STATUS_ERROR);
        }
    }

  files = &av[ssh_optind];
  num_files = ac - ssh_optind;

  /* Fetch default from ~/.ssh2/id_* (the first that we happen to get) */

#define ID_PREFIX "id_"
  
  if (ac == 1 && num_files == 0)
    {
      /* len includes '/' and '\0'*/
      len = strlen(ssh_user_dir(user)) + strlen(SSH_USER_DIR) + 2;
      ssh2dirname = ssh_xcalloc(len, sizeof(char));
      snprintf(ssh2dirname, len, "%s/%s", ssh_user_dir(user), SSH_USER_DIR);
      
      ssh2dir = opendir(ssh2dirname);

      if (!ssh2dir)
          ssh_fatal("Could not open directory \"%s\".", ssh2dirname);
          
      while ((cand = readdir(ssh2dir)) != NULL)
        {
          if (strlen(cand->d_name) >= strlen(ID_PREFIX) &&
              strncmp(cand->d_name, ID_PREFIX, strlen(ID_PREFIX)) == 0)
            {
              files = ssh_xcalloc(2, sizeof(char *));

              /* len includes '/' and '\0'*/
              len = strlen(ssh2dirname) + strlen(cand->d_name) + 2; 
                                                                      
              files[0] = ssh_xcalloc(len, sizeof(char));
              snprintf(files[0],len, "%s/%s", ssh2dirname, cand->d_name);
              ssh_xfree(ssh2dirname);
              num_files++;
              dynamic_array = TRUE;
              break;
            }
        }
      (void)closedir(ssh2dir);
    }
  
  signal(SIGPIPE, SIG_IGN);
  
  ssh_event_loop_initialize();
  
  ssh_agent_open(agent_open_callback, NULL);

  ssh_event_loop_run();
  ssh_event_loop_uninitialize();

  if (dynamic_array)
    {
      for(i = 0; i < num_files ; i++)
        {
          ssh_xfree(files[i]);
        }
      ssh_xfree(files);
    }
  
  ssh_user_free(user, FALSE);
  exit(EXIT_STATUS_OK);
}
